### ISSUE TITLE
BucketBatchSampler doesn't sort instances in descending order 

### ISSUE THREAD

**Issue by KwanWaiChung** [allenai/allennlp#4748]
In order to pass it to pack_padded_sequence, the sequences should be sorted in descending order. I am wondering whether it is the expected behavior or there are some parameters I didn't set correctly. 

Also, another question will be is there a way to retrieve the sequence length conveniently (i.e. tensor([3,4,])) in this case? My workaround is to compare each id to pad_id and take the sum. It's not complicated but it seems inconvenient that we have to always to do it ourselves,

A short example to reproduce. The shorter sequence is returned first as 0 is the pad id.
```python
from allennlp.data.tokenizers import Token
from allennlp.data.vocabulary import Vocabulary
from allennlp.data.fields import LabelField, TextField
from allennlp.data.dataset_readers.dataset_reader import AllennlpDataset
from allennlp.data.token_indexers import SingleIdTokenIndexer
from allennlp.data import Instance
from allennlp.data.samplers import BucketBatchSampler
from allennlp.data import PyTorchDataLoader

texts = [['a', 'b', 'c', 'd'], ['e'], ['f', 'g', 'h'], ['i', 'j']]
texts = [[Token(t) for t in tokens] for tokens in texts]
token_indexers = {'tokens': SingleIdTokenIndexer()}
instances = [Instance({'tokens': TextField(tokens, token_indexers)}) for tokens in texts]
dataset = AllennlpDataset(instances)
vocab = Vocabulary.from_instances(dataset)
dataset.index_with(vocab)

import random
random.seed(2)
batch_sampler = BucketBatchSampler(dataset, batch_size=2, sorting_keys=['tokens'])
data_loader = PyTorchDataLoader(dataset, batch_sampler=batch_sampler)
for batch in data_loader:
    print(batch)

>>> {'tokens': {'tokens': {'tokens': tensor([[7, 8, 9, 0],
        [2, 3, 4, 5]])}}}
```

**Comment by epwalsh** [comment]
Is there any reason you can't set `enforce_sorted=False` in the call to `pack_padded_sequence`? We also have a helper function called `sort_batch_by_length` in `allennlp.nn.util` which can be used to achieve the right sorting. You can see how this is used with `pack_padded_sequence` here: https://github.com/allenai/allennlp/blob/master/allennlp/modules/encoder_base.py

It does seems reasonable though to have a parameter in the `BucketBatchSampler` called `batch_sort` or `within_batch_sort` or something like that to handle this.

**Comment by KwanWaiChung** [comment]
> Is there any reason you can't set `enforce_sorted=False` in the call to `pack_padded_sequence`? We also have a helper function called `sort_batch_by_length` in `allennlp.nn.util` which can be used to achieve the right sorting. You can see how this is used with `pack_padded_sequence` here: https://github.com/allenai/allennlp/blob/master/allennlp/modules/encoder_base.py
> 
> It does seems reasonable though to have a parameter in the `BucketBatchSampler` called `batch_sort` or `within_batch_sort` or something like that to handle this.

Because sometimes I worked with decoupled models that do `pack_padded_sequence` inside `forward`, i.e. it expects the data to be sorted already. Secondly, I think even if I set `enforce_sorted=False`, pytorch will still have to do the sorting before further calculations.

For the example, I am wondering if we are using Allennlp, then how do we get the binary masking? Do we have to do manually?

**Comment by epwalsh** [comment]
> For the example, I am wondering if we are using Allennlp, then how do we get the binary masking? Do we have to do manually?

In many cases masks are created on the fly, but it's certainly possible to pre-compute masks as well in the form of an `ArrayField`.
